### DISCUSSION TITLE
How to use FrankenPHP with RequestStack

### DISCUSSION THREAD

**Discussion by stephanvierkant** [dunglas/symfony-docker#685] [Q&A]
I've used this template to get my webapplication running Docker. Almost everything works fine now, but I'm using a couple of [services with RequestStack](https://symfony.com/doc/current/service_container/request.html).

After some time of debugging, I found out this doesn't work great with FrankenPHP.

But how do I fix this? For now I disabled FrankenPHP, but I guess this should be doable.  I already found the ResetInterface, but I can't find good examples how to use it. Given the example, should `NewsletterManager` implement `Symfony\Contracts\Service\ResetInterface`?
  
update:
Let's say I've got something like this:

    final readonly class SomeService
    {
        public function __construct(private \Symfony\Component\HttpFoundation\RequestStack $requestStack)
        {
    
        }
        
        public function handleRequest() : Symfony\Component\HttpFoundation\Response
        {
            if ($this->requestStack->getCurrentRequest()->isXmlHttpRequest()) {
                return new Symfony\Component\HttpFoundation\JsonResponse();
            }
            
            return new Symfony\Component\HttpFoundation\Response('some html');
        }
    }

1. The controller gets a GET request, calls this service and returns the response. It serves 'some html'.
2. That page calls an AJAX request to the same endpoint, but now with a POST request. It still returns 'some html'

**Comment by 7-zete-7** [comment]
Hi @stephanvierkant!

Can you tell us more about the problem you encountered when using `RequestStack` and FrankenPHP?

`ResetInterface` is usually used for statefull services to reset their state at the start of processing a new request. When using CLI, CGI, FastCGI, etc., this interface is practically meaningless (since memory leaks and service states are eliminated when the handler process is restarted). In the case of using FrankenPHP, ReactPHP, AMPHP, etc., memory leaks and service states will remain because the handler process is not restarted. In this case, the service states, if any, must be cleared in the implementation of the `ResetInterface::reset()` method.

From the example of the `NewsletterManager` class - it does not require the implementation of `ResetInterface`. This class has no state.

Example of statefull service (bad practicle)
```php
class SomeValueStorage
{
    private ?string $someValue = null;

    public function __construct(
        private readonly RequestStack $requestStack,
    ) {
    }

    public function getSomeRequestedValue(): string
    {
        if (null !== $this->someValue) {
            return $this->someValue;
        }

        $request = $this->requestStack->getCurrentRequest();

        $someValue = $this->resolveSomeValue($request);
        $this->someValue = $someValue;

        return $someValue;
    }

    private function resolveSomeValue(Reuqest $request): string
    {
        // some heavy operations to resolve target value
    }
}
```

A class like this should be modified to not save its state. For this example, there are two options:

- Add a `ResetInterface` implementation
  ```diff
  @@ -1,4 +1,4 @@
  -class SomeValueStorage
  +class SomeValueStorage implements ResetInterface
   {
       private ?string $someValue = null;
  
  @@ -21,6 +21,11 @@
           return $someValue;
       }

  +    public function reset(): void
  +    {
  +        $this->someValue = null;
  +    }
  +
       private function resolveSomeValue(Reuqest $request): string
       {
           // some heavy operations to resolve target value
   }
  ```

- Store values ​​in request attributes
  ```diff
  @@ -1,7 +1,5 @@
   class SomeValueStorage
   {
  -    private ?string $someValue = null;
  -
       public function __construct(
           private readonly RequestStack $requestStack,
       ) {
  @@ -9,15 + 7,13 @@
  
       public function getSomeRequestedValue(): string
       {
  -        if (null !== $this->someValue) {
  -            return $this->someValue;
  -        }
  -
           $request = $this->requestStack->getCurrentRequest();

  +        if ('' === $someValue = $request->attributes->getString('some_value')) {
  +            $someValue = $this->resolveSomeValue($request);
  +            $request->attributes->set('some_value', $someValue);
  +        }
  +
  -        $someValue = $this->resolveSomeValue($request);
  -        $this->someValue = $someValue;
  -
           return $someValue;
       }

  ```

In both cases, the service loses the property that stores the state of the service, which makes this service stateless.

**Reply by stephanvierkant** [reply]
Thanks! I've updated my question. It looks like `$request` doesn't change because $requestStack is readonly and doesn't change between requests.

**Reply by 7-zete-7** [reply]
@stephanvierkant, thank you for clarifying the information in the question!

`readonly` does not allow changing the reference to the object, the object itself continues to change.

Is it possible that the request is sent via the Fetch API, and not via AJAX?
Is the `X-Requested-With: XMLHttpRequest` header passed for the test AJAX request?

**Reply by stephanvierkant** [reply]
Yes, `x-requested-with: XMLHttpRequest`.

**Reply by 7-zete-7** [reply]
@stephanvierkant, I tried to reproduce this behavior (see code below) - I was unable to reproduce this behavior.

Can you make a reproducer so that we can better understand what the problem might be?

<details><summary>Details</summary>
<p>

The code used to reproduce the problem:
```php
#[\Symfony\Component\HttpKernel\Attribute\AsController]
#[\Symfony\Component\Routing\Attribute\Route(path: '/foo', name: 'foo')]
class SomeController
{
    public function __construct(
        private readonly SomeService $someService,
    ) {
    }

    public function __invoke(): \Symfony\Component\HttpFoundation\Response
    {
        return $this->someService->handleRequest();
    }
}

class SomeService
{
    public function __construct(
        private readonly \Symfony\Component\HttpFoundation\RequestStack $requestStack,
    ) {
    }

    public function handleRequest(): \Symfony\Component\HttpFoundation\Response
    {
        $request = $this->requestStack->getCurrentRequest();

        if ($request->isMethod('POST')) {
            return new \Symfony\Component\HttpFoundation\JsonResponse([
                'is_ajax' => $request?->isXmlHttpRequest(),
            ]);
        }

        return new \Symfony\Component\HttpFoundation\Response(<<<'EOHTML'
        <div id="console"></div>

        <script>
          makeAjaxRequest(false, (res) => addConsoleLine(`AJAX without header: \${res}`));
          makeAjaxRequest(true, (res) => addConsoleLine(`AJAX with header: \${res}`));
          makeFetchRequest((res) => addConsoleLine(`Fetch: \${res}`));

          function addConsoleLine(message) {
            const el = document.createElement("p");
            el.textContent = message;
            document.getElementById("console").append(el);
          }

          function makeAjaxRequest(withHeader, callback) {
            const xhr = new XMLHttpRequest();

            xhr.open("POST", "/foo", true);

            if (withHeader) {
              xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
            }

            xhr.onreadystatechange = function () {
              if (this.readyState === 4 && this.status === 200) {
                callback(this.responseText);
              }
            };

            xhr.send();
          }

          function makeFetchRequest(callback) {
            fetch("/foo", {
              method: "POST",
            })
              .then((res) => res.text())
              .then((res) => callback(res));
          }
        </script>
        EOHTML);
    }
}
```

The response contained the following information:
```
AJAX without header: {"is_ajax":false}

AJAX with header: {"is_ajax":true}

Fetch: {"is_ajax":false}
```

The environment used:
- PHP version: 8.3.13
- FrankenPHP version: v1.2.5
- Symfony version: 6.4.14
- Server API: FrankenPHP (worker)

</p>
</details>

**Reply by stephanvierkant** [reply]
Thanks for your response! I'll try to create a reproducible repo.
